src/boot/efi/util.c

```diff
@@ -403,17 +408,46 @@ void sort_pointer_array(
 
         for (size_t i = 1; i < n_members; i++) {
                 size_t k;
-                void *entry = array[i];
+                uint8_t tmp[stride];
+                memcpy(tmp, array_element(array, stride, i), stride);
 
                 for (k = i; k > 0; k--) {
-                        if (compare(array[k - 1], entry) <= 0)
+                        if (compare(array_element(array, stride, k - 1), tmp) <= 0)
                                 break;
 
-                        array[k] = array[k - 1];
+                        memcpy(array_element(array, stride, k),
+                               array_element(array, stride, k - 1),
+                               stride);
                 }
 
-                array[k] = entry;
+                memcpy(array_element(array, stride, k), tmp, stride);
+        }
+}
+
+void *bsearch_array(
+                const void *key,
+                const void *array,
+                size_t stride,
+                size_t n_members,
+                compare_func_t compare) {
+        size_t left_idx = 0, mid_idx, right_idx = n_members;
+        const void *elem;
+
+        while (left_idx < right_idx) {
+                mid_idx = (left_idx + right_idx) / 2;
+                elem = array_element(array, stride, mid_idx);
+                int ret = compare(elem, key);
+                if (ret < 0)
+                        left_idx = mid_idx + 1;
+                else
+                        right_idx = mid_idx;
         }
+
+        elem = array_element(array, stride, left_idx);
+        if (compare(elem, key) == 0)
+                return (void *) elem;
+        else
+                return NULL;
```

Comment:
hmm, we usually want that the regular/successful codepath is the one that is not indented, and the exception/error code paths are the ones that are indented. That rule would suggest that this is written as:

```c
if (compare(elem, key) != 0)
        return NULL;

return (void*) elem;
```